### PR TITLE
[TIMOB-19040] :Replace MPMoviePlayerController in Ti.Media.VideoPlayer with AVPlayerViewController

### PR DESCRIPTION
**Pull request by vijaysingh-axway**
https://jira.appcelerator.org/browse/TIMOB-19040

### CODE REVIEW

**Comment by hansemannn** [comment]
Guess we need to move it to 7.0.0 because we have a couple of breaking changes that prevents us from putting it into non-major releases like 6.1.0.

**Comment by mattberg** [comment]
I am experiencing the issue described in:

https://jira.appcelerator.org/browse/TIMOB-20112

Which references the updates on this ticket. What is the best way to try and get these updates into my version of Titanium?

**Comment by hansemannn** [comment]
Hey there, it will be part of 7.0.0, but you can already cherry-pick this PR into a stable 6_1_X / 6_2_X branch and package your own SDK. Infos about packaging are in the readme of this repository, cherry-picking is described in many web-references. Good luck!

**Comment by vijaysingh-axway** [comment]
@hansemannn Resolved conflict. Please review. Thanks!

File: iphone/Classes/MediaModule.m [hunk 1]
```diff
@@ -458,6 +423,28 @@ - (void)setAudioSessionMode:(NSNumber *)mode
   }
 }
 
+- (NSNumber *)audioSessionMode
```
Comment (hansemannn) [inline iphone/Classes/MediaModule.m]: @vijaysingh-axway You added this API but deprecated it as well? Please clarify here.

File: iphone/Classes/TiMediaVideoPlayerProxy.h [hunk 2]
```diff
@@ -45,6 +49,7 @@
 @property (nonatomic, readwrite, assign) TiColor *backgroundColor;
 @property (nonatomic, readonly) NSNumber *playing;
 @property (nonatomic, copy) NSNumber *volume;
+@property (nonatomic, readwrite, assign) id pictureInPictureEnabled;
```
Comment (hansemannn) [inline iphone/Classes/TiMediaVideoPlayerProxy.h]: Should be `NSNumber` if public API. `BOOL` otherwise.

**Review by hansemannn — CHANGES_REQUESTED** [review]

File: iphone/Classes/MediaModule.m (hunk 1 above)
Comment (vijaysingh-axway) [reply iphone/Classes/MediaModule.m]: @hansemannn Good catch! This method was added by you in your initial change for this ticket. I think it is not needed here as audioSessionCategory is already there. Even variables AUDIO_SESSION_MODE_* are also not defined there. So removed this method. Thanks.

File: iphone/Classes/TiMediaVideoPlayerProxy.h (hunk 2 above)
Comment (vijaysingh-axway) [reply iphone/Classes/TiMediaVideoPlayerProxy.h]: Made it NSNumber.

File: iphone/Classes/TiMediaVideoPlayerProxy.m [hunk 3]
```diff
@@ -457,39 +375,84 @@ - (void)setVolume:(NSNumber *)newVolume
...
+{
+  ENSURE_UI_THREAD_1_ARG(time);
+  if (movie != nil) {
+    double ourTime = [TiUtils doubleValue:time];
+    if (ourTime > 0 || isnan(ourTime)) {
+      ourTime /= 1000.0f; // convert from milliseconds to seconds
+
+      // Handle both setting this value on running videos and on creation
+      if ([[movie player] status] == AVPlayerStatusReadyToPlay) {
+        [[movie player] seekToTime:CMTimeMake(ourTime, 1)];
```
Comment (hansemannn) [inline iphone/Classes/TiMediaVideoPlayerProxy.m]: This does not work. Try to use
```objc
[[movie player] seekToTime:CMTimeMake(ourTime, 1) toleranceBefore:kCMTimeZero toleranceAfter:kCMTimeZero];
```
instead.

**Review by hansemannn — CHANGES_REQUESTED** [review]
Some more issues I found during testing. Please also remove the `demos/KitchenSink/Resources/another.mp4` file and add an example to the `kitchensink-v2` instead.

File: iphone/Classes/TiMediaVideoPlayerProxy.h [hunk 4]
```diff
@@ -13,9 +13,15 @@
...
 
+enum {
+  VideoTimeOptionNearestKeyFrame = 0,
+  VideoTimeOptionExact,
+};
+
 @interface TiMediaVideoPlayerProxy : TiViewProxy {
   @protected
   AVPlayerViewController *movie;
+  MPMoviePlayerViewController *cont;
```
Comment (hansemannn) [inline iphone/Classes/TiMediaVideoPlayerProxy.h]: Why did you add the deprecated API again here?

File: iphone/Classes/TiMediaVideoPlayerProxy.h [hunk 5]
```diff
@@ -13,9 +13,15 @@
 #import <AVKit/AVKit.h>
 #import <MediaPlayer/MediaPlayer.h>
 
+enum {
+  VideoTimeOptionNearestKeyFrame = 0,
+  VideoTimeOptionExact,
+};
```
Comment (hansemannn) [inline iphone/Classes/TiMediaVideoPlayerProxy.h]: Use NS_Enums.

**Review by hansemannn — CHANGES_REQUESTED** [review]

**Comment by hansemannn** [comment]
Please also add unit-tests here, there are none so far. The tests can be pretty straight forward with the test-suite you provided earlier.

File: iphone/Classes/MediaModule.h [hunk 6]
```diff
@@ -180,13 +179,18 @@
...
-@property (nonatomic, readonly) NSNumber *VIDEO_CONTROL_FULLSCREEN; // Controls for fullscreen playback
-
-@property (nonatomic, readonly) NSNumber *VIDEO_MEDIA_TYPE_NONE;
-@property (nonatomic, readonly) NSNumber *VIDEO_MEDIA_TYPE_VIDEO;
-@property (nonatomic, readonly) NSNumber *VIDEO_MEDIA_TYPE_AUDIO;
+/* REMOVED since iOS9
+@property(nonatomic,readonly) NSNumber* VIDEO_CONTROL_NONE;			// No controls
+@property(nonatomic,readonly) NSNumber* VIDEO_CONTROL_EMBEDDED;		// Controls for an embedded view
+@property(nonatomic,readonly) NSNumber* VIDEO_CONTROL_FULLSCREEN;	// Controls for fullscreen playback
+*/
```
Comment (hansemannn) [inline iphone/Classes/MediaModule.h]: Remove this comment.

File: apidoc/Titanium/Media/Media.yml [hunk 7]
```diff
@@ -1176,6 +1176,10 @@ properties:
     type: Number
     permission: read-only
   - name: VIDEO_CONTROL_DEFAULT
+    deprecated:
+        since: '7.0.0'
+        notes: This property has been removed for iOS in Titanium SDK 7.0.0 as of the official deprecation by Apple.
+        removed: '7.0.0'
```
Comment (hansemannn) [inline apidoc/Titanium/Media/Media.yml]: Are single quotes ok in the docs? We usually use `"`. Please validate the docs locally again. cc @jawa9000

File: apidoc/Titanium/Media/Media.yml [hunk 8]
```diff
@@ -1265,17 +1277,30 @@ properties:
     type: Number
     permission: read-only
     platforms: [android, iphone, ipad]
+  - name: VIDEO_LOAD_STATE_FAILED
+    summary: Indicates that the player can no longer play AVPlayerItem instances because of an error.
```
Comment (hansemannn) [inline apidoc/Titanium/Media/Media.yml]: Titanium users have no idea what a `AVPlayerItem` is. Use "media items" instead.

File: apidoc/Titanium/Media/Media.yml [hunk 9]
```diff
@@ -1361,30 +1414,68 @@ properties:
...
   - name: VIDEO_SCALING_NONE
+    deprecated:
+        since: '7.0.0'
+        notes: This property has been removed for iOS in Titanium SDK 7.0.0 as of the official deprecation by Apple.
+        removed: '7.0.0'
     summary: Video scaling is disabled.
     type: Number
     permission: read-only
+  - name: VIDEO_SCALING_RESIZE
+    summary: Specifies that the video should be stretched to fill the layer’s bounds.
```
Comment (hansemannn) [inline apidoc/Titanium/Media/Media.yml]: That `’` probably fails validation.

**Review by hansemannn — CHANGES_REQUESTED** [review]

File: apidoc/Titanium/Media/Media.yml (hunk 7 above)
Comment (vijaysingh-axway) [reply apidoc/Titanium/Media/Media.yml]: Very few places single quotes are there. I updated to ".

File: apidoc/Titanium/Media/Media.yml (hunk 9 above)
Comment (vijaysingh-axway) [reply apidoc/Titanium/Media/Media.yml]: At other places in doc also  '  is used. So I think no problem with this.

**Comment by hansemannn** [comment]
@vijaysingh-axway Can you please re-publish your branch? It was accidentally merged and I did not test the latest changes so far. Thanks!
